File: trail5/treeset 기본/treeset-basic.py
n = int(input())
command = []
x = []

for _ in range(n):
    line = input().split()
    command.append(line[0])
    if line[0] in ["add", "remove", "find", "lower_bound", "upper_bound"]:
        x.append(int(line[1]))
    else:
        x.append(0)
"""
add x : 숫자 x를 treeset에 추가합니다. 중복되는 경우 무시합니다.
remove x : 숫자 x를 treeset에서 제거합니다. 잘못된 입력은 주어지지 않습니다.
find x : 숫자 x가 treeset에 있는지를 판단합니다. 있다면 true, 없다면 false를 출력합니다.
lower_bound x : treeset에서 x보다 같거나 큰 최초의 숫자를 출력합니다. 만약 없다면 None을 출력합니다.
upper_bound x : treeset에서 x보다 큰 최초의 숫자를 출력합니다. 만약 없다면 None을 출력합니다.
largest : treeset에서 가장 큰 숫자를 출력합니다. 만약 treeset이 비어있다면 None을 출력합니다.
smallest : treeset에서 가장 작은 숫자를 출력합니다. 만약 treeset이 비어있다면 None을 출력합니다.
"""
from sortedcontainers import SortedSet
ss = SortedSet()

for i in range(n):
    cmd, num = command[i], x[i]
    if cmd == 'add':
        ss.add(num)
    elif cmd =='remove':
        ss.remove(num)
    elif cmd == 'find':
        print('true' if num in ss else 'false')
    elif cmd == 'lower_bound':
        idx = ss.bisect_left(num)
        if idx == len(ss):
            print('None')
            continue
        print(ss[idx])
    elif cmd == 'upper_bound':
        idx = ss.bisect_right(num)
        if idx == len(ss):
            print('None')
            continue
        print(ss[idx])
    elif cmd == 'largest':
        if not ss:
            print('None')
            continue
        print(ss[-1])
    else:
        if not ss:
            print('None')
            continue
        print(ss[0])

File: trail5/treeset 기본/treeset-basic.java
import java.util.*;
import java.io.*;

class Query {
    String cmd;
    int x;

    public Query(String cmd, int x) {
        this.cmd = cmd;
        this.x = x;
    }
}

public class Main {
    static int N, x;
    static Integer ans;
    static String cmd;
    static List<Query> qlst = new ArrayList<>();
    static TreeSet<Integer> set = new TreeSet<>();

    static void init() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());

        for (int q=0; q<N; q++) {
            st = new StringTokenizer(br.readLine());
            cmd = st.nextToken();

            if (st.hasMoreTokens()) {
                x = Integer.parseInt(st.nextToken());
            }
            else {
                x = 0;
            }
            qlst.add(new Query(cmd, x));
        }
    }

    public static void main(String[] args) throws IOException {
        init();

        for (Query q : qlst) {
            // System.out.println(set);

            if (q.cmd.equals("add")) {
                set.add(q.x);
            }

            else if (q.cmd.equals("remove")) {
                set.remove(q.x);
            }

            else if (q.cmd.equals("find")) {
                if (set.contains(q.x)) {
                    System.out.println("true");
                } else {
                    System.out.println("false");
                }
            }

            else if (q.cmd.equals("lower_bound")) {
                ans = set.ceiling(q.x);
                if (ans == null) {
                    System.out.println("None");
                } else {
                    System.out.println(ans);
                }
            }

            else if (q.cmd.equals("upper_bound")) {
                ans = set.higher(q.x);
                if (ans == null) {
                    System.out.println("None");
                } else {
                    System.out.println(ans);
                }
            }

            else if (q.cmd.equals("largest")) {
                if (set.size() == 0) {
                    System.out.println("None");
                    continue;
                }
                ans = set.last();
                System.out.println(ans);
            }

            else {
                if (set.size() == 0) {
                    System.out.println("None");
                    continue;
                }
                ans = set.first();
                System.out.println(ans);
            }
        }
    }
}